accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,802 @@
+simplify the spec, there is a single procedure used for all kinds of
+collections. This recursive procedure builds up either a `result` sequence of
+either values or key-value pair map entries. Then a final step handles those
+appropropriately for the given collection type.
+
+### To evaluate a collection `element`:
+
+1.  If `element` is an expression element:
+
+    1.  Evaluate the element's expression and append it to `result`.
+
+1.  Else, `element` is a `mapEntry` `keyExpression: valueExpression`:
+
+    1.  Evaluate `keyExpression` to a value `key`.
+
+    1.  Evaluate `valueExpression` to a value `value`.
+
+    1.  Append `key: value` to `result`.
+
+1.  Else, if `element` is a spread element:
+
+    1.  Evaluate the spread expression to a value `spread`.
+
+    1.  If `entry` is null-aware and `spread` is null, continue to the next
+        element in the literal.
+
+    1.  If the collection is a map, evaluate `spread.entries.iterator` to a
+        value `iterator`. Otherwise, evaluate `spread.iterator` to `iterator`.
+        *This will deliberately throw an exception if `spread` is `null` and
+        `element` is not null-aware.*
+
+    1.  Loop:
+
+        1.  If `iterator.moveNext()` returns `false`, exit the loop.
+
+        1.  Evaluate `iterator.current` and append it to `result`. *This will be
+            a MapEntry in a map literal, or any object for a list or set
+            literal.*
+
+1.  Else, if `element` is an `if` element:
+
+    1.  Evaluate the condition expression to a value `condition`.
+
+    1.  If the boolean conversion of `condition` is `true`:
+
+        1.  Evaluate the "then" element using this procedure.
+
+    1.  Else, if there is an "else" element of the `if`:
+
+        1.  Evaluate the "else" element using this procedure.
+
+1.  Else, if `element` is a synchronous `for-in` element:
+
+    1.  Evaluate the iterator expression to a value `sequence`.
+
+    1.  Evaluate `sequence.iterator` to a value `iterator`.
+
+    1.  Loop:
+
+        1.  If the boolean conversion of `iterator.moveNext()` does not return
+            `true`, exit the loop.
+
+        1.  If the `for-in` element declares a variable, create a new namespace
+            and a fresh `variable` for it. Otherwise, use the existing
+            `variable` it refers to.
+
+        1.  Evaluate `iterator.current` and bind it to `variable`.
+
+        1.  Evaluate the body element using this procedure in the scope of
+            `variable`.
+
+        1.  If the `for-in` element declares a variable, discard the namespace
+            created for it.
+
+1.  Else, if `element` is an asynchronous `await for-in` element:
+
+    1.  Evaluate the stream expression to a value `stream`. It is a dynamic
+        error if `stream` is not an instance of a class that implements
+        `Stream`.
+
+    1.  Create a new `Future`, `streamDone`.
+
+    1.  Evaluate `await streamDone`.
```

Comment:
That doesn't work. An "evaluation" doesn't stop until it throws or yields a value, so the steps below will not happen.

Set up the stream listening first. Then write that you wait for the *streamDone* future to complete.
(Or don't use a *streamDone* future, and just say that evaluation throws errors and stack, or that evaluation completes normally).

> 1. Listen on *stream* and take the following actions on events:
>     1. On a data event with value *v*, bind *v* to *variable*, then
>         evaluate the body using this algorithm. (plus scope things).
>         If evaluation of the body throws an error *e* and stack trace *s*, 
>         then stop listening on the stream by calling the `cancel` method
>         of the corresponding stream subscription, which returns a future *f*.
>         Then wait for *f* to complete. If *f* completes with an error *e2* and
>         stack trace *s2*, then complete *streamDone* with error *e2* and
>         stack trace *s2*. Otherwise complete *streamDone* with error *e*
>         and stack trace *s*.
>         If evaluation of the body element completes successfully,
>         resume the stream subscription if it has been paused.
>     2. On an error event with error *e* and stack trace *s*,
>         stop listening on the stream by calling the `cancel` method
>         of the corresponding stream subscription, which returns a future *f*.
>         Then wait for *f* to complete. If *f* completes with an error *e2* and
>         stack trace *s2*, then complete *streamDone* with error *e2* and
>         stack trace *s2*.
>     3. On a done event, complete *streamDone* with the value `null`.
>
>  Evaluation of the `for-in` element completes when *streamDone* completes. 
> If *streamDone* completes with an error *e* and stack trace *s*, 
> then evaluation of the element throws *e* with stack trace *s*, 
> otherwise the evaluation completes successfully.

or 

> 1. Pause any surrounding `await-for` loops.
> 1. Listen on *stream* and take the following actions on events:
>     1. On a data event with value *v*, bind *v* to *variable*, then
>         evaluate the body element using this algorithm. (plus scope things).
>         If evaluation of the body throws an error *e* and stack trace *s*, 
>         then stop listening on the stream by calling the `cancel` method
>         of the corresponding stream subscription, which returns a future *f*.
>         Then wait for *f* to complete. If *f* completes with an error *e2* and
>         stack trace *s2*, then evaluation of the `for-in` elements 
>         throws error *e2* and stack trace *s2*. 
>         Otherwise the evaluationt hrows error *e* and stack trace *s*.
>         If evaluation of the body element completes successfully,
>         resume the stream subscription if it has been paused.
>     2. On an error event with error *e* and stack trace *s*,
>         stop listening on the stream by calling the `cancel` method
>         of the corresponding stream subscription, which returns a future *f*.
>         Then wait for *f* to complete. If *f* completes with an error *e2* and
>         stack trace *s2*, then evaluation of the `for-in` elements 
>         throws error *e2* and stack trace *s2*. 
>     3. On a done event, evaluation of the `for-in` element completes successfully.
> 2. Evaluation of the `for-in` element doesn't complete until the above event handlers say so.

Reply:
Done. Let me know if I got it right.